web/src/message_list_view.js

```diff
@@ -1533,5 +1533,23 @@ export class MessageListView {
         const today = new Date();
         const rendered_date = timerender.render_date(time, undefined, today);
         $sticky_header.find(".recipient_row_date").html(rendered_date);
+
+        // Date is always visible for date separator headers in a message list.
+        // This can cause two headers with same date visible at the top of message
+        // feed. To avoid this, we hide the date on the date separator until next user
+        // scroll if the date separator is placed before a sticky header.
+        $(".hide-date-separator-header").removeClass("hide-date-separator-header");
+        if ($sticky_header.find(".recipient_row_date.hide-date").length) {
+            const $prev_recipient_row = $sticky_header
+                .closest(".recipient_row")
+                .prev(".recipient_row");
+            if (!$prev_recipient_row.length) {
+                return;
+            }
+            const $prev_header_date_row = $prev_recipient_row.find(".recipient_row_date");
+            if (!$prev_header_date_row.hasClass("hide-date")) {
+                $prev_header_date_row.addClass("hide-date-separator-header");
+            }
+        }
```

Comment:
Can you explain the logic here a bit more? I'm not sure I understand what either `if` statement is checking for, nor why this is the right check to do.

Reply:
Added more comments in the PR.